### PR TITLE
Converted to TypeScript

### PR DESCRIPTION
**Pull request by ChrisMBarr**
I wanted to use this library in a webapp project of mine but I was frustrated because it was not at all working (see issue #34 ), and my project is written in TypeScript and this has no typedefs associated with it so even if it did work I would not be able to get any intellisense or compiler feedback.

Since this is not a huge library, I was able to convert it to TypeScript!  It still outputs JS as before, but I have made several improvements to the code as well, but there are **no functionality changes**!

### PR Highlights
* The package name `songpro` now matches the casing when importing it into a project. See the diff on the ReadMe.md file that reflects this.
* Added more tests, and organized the existing tests. Now there is 100% code coverage.
* There were a lot of areas that used `null` or `undefined` (sometimes both) as well as the real value. This is now less messy.
* In general adding type definitions on everything make the code safer and easier to maintain and improve upon in the future
* Only 2 classes now, the rest are TypeScript interfaces. Now fewer objects need to be `new`ed up, and there is less code being output to JS. Only the `SongPro` class and the TypeScript interfaces are exported.
* The dependency on LoDash is removed. It only used 2 functions anyway, which can both be done without a library
* The dependencies for Babel and Webpack are removed as well.
* Large methods have been split out into smaller more focused ones. Every method now has a cyclomatic complexity of 8 or less, which is now enforced by an ESLint rule
* More ESLint rules have been added, some specific to TypeScript
* Now uses a standard JS module, so this package can easily be added to both Node.js apps as before, plus web apps

### Summary
Basically the code has been converted and improved in many ways, but it still works exactly the same as before.

I'd love to see this get merged in pretty soon as I want to use this package on my project, but FYI I will probably just go ahead and publish my own fork of it on NPM so I can use it now.

Please let me know if you have any questions or concerns about my changes!

### Code Coverage & Tests
code-coverage

### Intellisense working in a TypeScript project
songpro-intellisense

### CODE REVIEW

**Comment by spilth** [comment]
Hi there. I've been sick the last few days so I'm getting to this PR a little late. The JS library was not in particularly good shape as I mostly rely on the Ruby version currently.

It is a large PR and a definitely a shift with switching to TypeScript, so it might take me some time to look through it and accept it. But I wanted to reply to let you know I've seen your PR.

**Comment by spilth** [comment]
@FiniteLooper Do you think it would make more sense to have this code in a repository named `songpro-typescript`? I'm happy to make that and give you access to it if so.

**Comment by ChrisMBarr** [comment]
Well the output is still javascript in the end, so in that way it's really the same end result.  Hmm, I don't know why it's failing the build.  It can't locate the module from the tests file.  IDK why that would happen... I hate to say "it works on my machine" though.

If you'd prefer to keep this a separate repo that's fine and I don't mind keeping this on my own page as a fork.  if so I'll probably rename it though to prevent name conflicts.  It's up to you really!

**Comment by spilth** [comment]
Alright, let's keep it here then. I gave you Write access to the repo, if that helps you debug the CI process.

**Comment by spilth** [comment]
I get the following when I checkout your repo and run `rpm run build`:

```
> jest --coverage

 FAIL  src/SongPro.test.ts
  ● Test suite failed to run

    src/SongPro.test.ts:1:25 - error TS1149: File name '/Users/brian/workspace/temp/songpro-javascript/src/songpro.ts' differs from already included file name '/Users/brian/workspace/temp/songpro-javascript/src/SongPro.ts' only in casing.
      The file is in the program because:
        Root file specified for compilation
        Imported via "./songpro" from file '/Users/brian/workspace/temp/songpro-javascript/src/SongPro.test.ts'

    1 import { SongPro } from "./songpro";
                              ~~~~~~~~~~~

----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |       0 |        0 |       0 |       0 |                   
----------|---------|----------|---------|---------|-------------------
Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        1.052 s
Ran all test suites.
```

**Comment by ChrisMBarr** [comment]
try running `npm run build-test` first and then run `npm run test` or `npm run build`.  If that fixes it then that one script just needs to be added to the `build` script

**Comment by ChrisMBarr** [comment]
And about that casing error, I did re-case the file name to be all lowercase.  If that is giving you a problem open the `tsconfig.ts` and disable `forceConsistentCasingInFileNames`.  I'm surprised it didn't update that when you checked it out, but I guess git doesn't deal with file name casing usually

**Comment by spilth** [comment]
The build is managed in `.github/workflows/build.yml`, so you have the ability to update that in your PR.

**Comment by FiniteLooper** [comment]
Hmm, it's not letting me push with changes to `build.yml`

```
Pushing to https://github.com/FiniteLooper/songpro-javascript
To https://github.com/FiniteLooper/songpro-javascript
 ! [remote rejected] main -> main (refusing to allow an OAuth App to create or update workflow `.github/workflows/build.yml` without `workflow` scope)
error: failed to push some refs to 'https://github.com/FiniteLooper/songpro-javascript'
```

**Comment by ChrisMBarr** [comment]
ugh. I resort to StackOverflow and call it a night
https://stackoverflow.com/questions/76341013/github-actions-fail-when-running-ts-jest-tests-with-cannot-find-module-or-i

**Comment by spilth** [comment]
Since the casing thing came up, is this possibly related? https://stackoverflow.com/a/64359597/8252

**Comment by ChrisMBarr** [comment]
HA! that was it! Well at least for a few of the builds.

Since Node 16 and 18 passed, but 15 failed and the rest didn't run.  My guess is that Node <=15 can't run the JS that is being output.  TypeScript is outputting ES2022, and a quick google shows Node 15 doesn't support that.  But that's an easy change!

**Review by spilth — APPROVED** [review]

**Comment by ChrisMBarr** [comment]
Man I can't believe all those problems were only due to file name casing!  Thanks for that find, that was the key!

FYI: the reason behind me doing that was because the NPM package is named `songpro` but before you had to import it like `require('SongPro')` which was confusing because usually you just import by the package name.  Now it's all consistent.
